### PR TITLE
Flag PATCH users/self route as Deprecated

### LINKED ISSUE #2155: The PATCH /users/self route is ambiguous
### Issue Description

The Patch /users/self routes, doesn't allow the `developers` to update their profile data.
By `developers` we mean those users who are in the discord server and whom user Details is not incomplete. 
For user to update their profile data we suggest them to use Profile service.

### Expected Behaviour

If a request is made to the route `PATCH /users/self` with some data in the `body: partial<UserObject>` it is the expectation that it  should update user data with all that is being sent in the body  

### Current Behaviour

If we send a request with object 
```js
body: {
      username:string
}
```
it return can't update the username twice -> expected 
```js
body:{
   roles:{
         super_user: true/false
  }
}
```
it return can't modify roles --> expected

For other properties we don't allow users to update their profile data, if the user is a `developer`, `(in_discord==true && !userDetailsIncomplete)`. 
this route is generally used to add details to user once they are signed up or has their userDetailsIncomplete property set to true.

#### Flow
image

### Reproducibility

 - [x] This issue is reproducible
 - [ ] This issue is not reproducible

### Steps to Reproduce
send the objects mentioned in current behaviour section on `PATCH /users/self`.

### Severity/Priority
 - [ ] Critical
 - [ ] High
 - [ ] Medium
 - [x] Low

### Checklist

- [x] I have read and followed the project's code of conduct.
- [x] I have searched for similar issues before creating this one.
- [x] I have provided all the necessary information to understand and reproduce the issue.
- [x] I am willing to contribute to the resolution of this issue.

### PR DESCRIPTION
**Pull request by shubhdevelop**
Date: 06/10/2024

Developer Name: Shubham Raj

---

## Issue Ticket Number
Closes #2155 

The above mentioned issue was related to improving the overall performance of the patch users/self, thought we are either way going to deprecate all the `users/self` route in accordance with issue #2126, So there is no point wasting effort on something that we are going to deprecate, 

I'm adding comments, above the controller and the endpoint, as well as update the api contracts documentation

## TODO
- [ ] Implementing the new route replacing the current PATCH "/user/self" route as well as it's controller
- #2126 
- **[vikas](https://github.com/vikasosmium)**  is working on the issue. 

### Documentation Updated?

[This PR updates the API Contracts](https://github.com/Real-Dev-Squad/website-api-contracts/pull/193)
**Note:**
- [x] Documentation merged
-------

- [ ] No
- [x] Yes

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [x] Yes
- [ ] No

### Development Tested?

Adds an handy warning in the IDE

image

- [x] Yes
- [ ] No

## Test Coverage
Didn't add any test coverage only modified the previously written one and they all are passing

## Additional Notes:
changes done in accordance with solution 3
https://docs.google.com/document/d/1ub9ryFs4qq9m9m5jdOxFAx9VvBPMpOn4-_jDTBaNkJY/edit?tab=t.0

### CODE REVIEW

File: test/integration/users.test.js [hunk 1]
```diff
@@ -242,7 +243,7 @@ describe("Users", function () {
           },
         });
 
-      expect(updateRolesResponse).to.have.status(204);
+      expect(updateRolesResponse).to.have.status(200);
```
Comment (iamitprakash) [inline test/integration/users.test.js:246]: why you are checking for 200 status code only, what about others codes

**Review by iamitprakash — CHANGES_REQUESTED** [review]
comment added

File: controllers/users.js [hunk 2]
```diff
@@ -388,6 +388,7 @@ const getSelfDetails = async (req, res) => {
 
 /**
  * Update the user
+ * @deprecated [SCHEDULED] This Controller will be removed in the near Future, New controller TBD.
```
Comment (yesyash) [inline controllers/users.js:391]: - will you be adding the alternative to this? if not can you please mention the user id of the person you know?

Comment (shubhdevelop) [reply controllers/users.js:391]: no I'll not be adding the alternatives, [vikas](https://github.com/vikasosmium) is working on deprecating /users/self routes

Comment (yesyash) [reply controllers/users.js:391]: @shubhdevelop please add a todo comment with vikas github id stating he would be adding this alternative

Comment (shubhdevelop) [reply controllers/users.js:391]: Added

Comment (shubhdevelop) [reply controllers/users.js:391]: @yesyash Added, todo comments can you please review & approve of the changes

Comment (sahsisunny) [reply controllers/users.js:391]: I think mentioning the issue ticket here is sufficient.

Comment (shubhdevelop) [reply controllers/users.js:391]: I had added, as asked by yash

**Review by yesyash — APPROVED** [review]

**Review by vinit717 — APPROVED** [review]

**Review by Achintya-Chatterjee — APPROVED** [review]

**Comment by prakashchoudhary07** [comment]
@shubhdevelop Why was this closed?

**Comment by shubhdevelop** [comment]
got a comment about this from Tejas sir, and work on the new api is on & vikas is doing it

**Pull request closed without merging**